third_party/move/move-vm/types/src/utils/union_find.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#[derive(Debug)]
+pub struct UnionFind {
+    parent: Vec<u32>,
+    rank: Vec<u8>,
+    num_sets: usize,
+    capacity: usize,
+}
+
+impl UnionFind {
+    /// Creates a new UnionFind instance with `num_elements` and `capacity`
+    pub fn new_with_size_and_capacity(
+        num_elements: usize,
+        capacity: usize,
+    ) -> Result<Self, &'static str> {
+        if num_elements > std::u32::MAX as usize {
+            return Err("num_elements larger than std::u32::MAX");
+        }
+        if capacity > std::u32::MAX as usize {
+            return Err("capacity larger than std::u32::MAX");
+        }
+        if num_elements > capacity {
+            return Err("num_elements larger than given capacity");
+        }
+
+        let parent = (0..num_elements as u32).collect::<Vec<u32>>();
+        let rank = vec![0u8; num_elements];
+        Ok(Self {
+            parent,
+            rank,
+            num_sets: num_elements,
+            capacity,
+        })
+    }
+
+    /// Creates a new UnionFind instance with given `capacity` and 0 elements
+    pub fn new_with_capacity(capacity: usize) -> Result<Self, &'static str> {
+        Self::new_with_size_and_capacity(0, capacity)
+    }
+
+    /// Creates a new UnionFind instance with `num_elements` and maximal capacity.
+    pub fn new_with_size(num_elements: usize) -> Result<Self, &'static str> {
+        Self::new_with_size_and_capacity(num_elements, std::u32::MAX as usize)
+    }
+
+    /// Creates a new UnionFind instance with `0` elements and maximal capacity.
+    pub fn new() -> Result<Self, &'static str> {
+        Self::new_with_size_and_capacity(0, std::u32::MAX as usize)
+    }
+
+    // Finds the representative of the set containing `id` without bound checking
+    fn find_set_internal(&mut self, id: u32) -> u32 {
+        let mut i = id;
+        let mut n = id;
+
+        while n != self.parent[n as usize] {
+            n = self.parent[n as usize];
+        }
+
+        while n != i {
+            i = std::mem::replace(&mut self.parent[i as usize], n);
+        }
+
+        n
+    }
+
+    /// Finds the representative of the set containing `id`
+    pub fn find_set(&mut self, id: u32) -> Result<u32, &'static str> {
+        if (id as usize) >= self.size() {
+            return Err("Element out of bounds");
+        }
+
+        Ok(self.find_set_internal(id))
+    }
+
+    // Joins sets represented by `x` and `y` (both have to be set representative)
+    pub fn join_sets_internal(&mut self, x: u32, y: u32) -> u32 {
+        if x == y {
+            return x;
+        }
+
+        // Merging two different sets
+        self.num_sets -= 1;
+
+        if self.rank[x as usize] >= self.rank[y as usize] {
```

Comment:
This code might be more readable if you started with:
```
   let (xidx, yidx) = (x as usize, y as usize);
```
and then use those instead of `[xy] as usize` everywhere.

Reply:
That would require downcasting xidx/yidx to u32. I prefer not to downcast, even though it is safe here.

Reply:
No, you could also keep x and y:
```
    let (xidx, yidx) = (x as usize, y as usize);
    if self.rank[xidx] >= self.rank[yidx] {
        self.parent[yidx] = x;
        if self.rank[xidx] == self.rank[yidx] {
            self.rank[xidx] += 1;
        }
        return x;
    }
    self.parent[xidx] = y;
    y
}
```
but actually, I think just manually CSEing the rank reads makes it cleaner, as well as a more parallel construction:
```
    let rank_x = self.rank[x as usize];
    let rank_y = self.rank[y as usize];
    if rank_x >= rank_y {
        if rank_x == rank_y {
            self.rank[x as usize] = rank_x + 1;
        };
        self.parent[y as usize] = x;
        x
    } else {
        self.parent[x as usize] = y;
        y
    }
```
Time for me to go to sleep, obviously.  This doesn't really matter.